CSS Custom Highlights WPT painting-invalidation 003 & 006, painting-iframe 006 failing due to not updating positions correctly

jesxilin opened the pull request:
#### 0ef0801e78b627d421da475271e6f9385e18af4c
<pre>
CSS Custom Highlights WPT painting-invalidation 003 &amp; 006, painting-iframe 006 failing due to not updating positions correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=259521">https://bugs.webkit.org/show_bug.cgi?id=259521</a>
rdar://112907577

Reviewed by Richard Robinson and Ryosuke Niwa.

Added check for live range in Document::collectRangeDataFromRegister because
we do not want to skip if a live range is already set.
Live ranges are able to change.
Took out null checks in Document::updateHighlightPositions to support that
live ranges can change and there are already existing null checks before setting positions.
Updated TestExpectations due to three passing tests.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::collectRangeDataFromRegister):
(WebCore::Document::updateHighlightPositions):
* Source/WebCore/dom/Range.h:
* Source/WebCore/dom/StaticRange.h:

Canonical link: <a href="https://commits.webkit.org/266349@main">https://commits.webkit.org/266349@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/58317c481f31acb1ddc823180a61e56093815e70

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13595 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13909 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14242 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15331 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/12914 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13676 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16417 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13989 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15607 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13761 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14404 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11511 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16030 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11692 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/12266 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/19302 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12767 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12434 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15639 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12946 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10838 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12216 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3312 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16546 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12789 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -3015,10 +3016,8 @@ void Document::updateHighlightPositions()
-            if (rangeData->endPosition().isNull())
-                endPosition = visibleSelection.visibleEnd().deepEquivalent();
+            startPosition = visibleSelection.visibleStart().deepEquivalent();
+            endPosition = visibleSelection.visibleEnd().deepEquivalent();
```
These can be combined fwiw

nt1m replied:
```
            auto startPosition = visibleSelection.visibleStart().deepEquivalent();
            auto endPosition = visibleSelection.visibleEnd().deepEquivalent();
```

jesxilin replied:
On it!

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2988,7 +2988,8 @@ void Document::collectRangeDataFromRegister(Vector<WeakPtr<HighlightRangeData>>&
     for (auto& highlight : highlightRegister.map()) {
         for (auto& rangeData : highlight.value->rangesData()) {
-            if (rangeData->startPosition().isNotNull() && rangeData->endPosition().isNotNull())
+            // FIXME: Change to a flag that checks if liveRange was changed rather than if liveRange for optimization.
```
```suggestion
            // FIXME: For live ranges, we can optimize by only performing this when the range changed.
```

It took me a bit to understand the current comment

jesxilin replied:
Will change!

rniwa commented on Source/WebCore/dom/Range.h:
```diff
@@ -52,6 +52,8 @@ class Range final : public AbstractRange, public CanMakeWeakPtr<Range> {
     bool collapsed() const final { return m_start == m_end; }
     WEBCORE_EXPORT Node* commonAncestorContainer() const;
 
+    bool isLiveRange() const final { return true; }
```
Why are we making this public?? It's a public member of `AbstractRange` so we shouldn't have to make this public here. If any code is relying on this, it means we're calling `isLiveRange` on `Range` object, which is redundant.

jesxilin replied:
Yep, I will change it!

aprotyas commented on Source/WebCore/dom/StaticRange.h:
```diff
@@ -67,6 +67,7 @@ class StaticRange final : public AbstractRange, public SimpleRange {
     StaticRange(SimpleRange&&);
 
     bool isLiveRange() const final { return false; }
+
```
nit: Remove the blank line plz

jesxilin replied:
will do!

rniwa commented on Source/WebCore/dom/Range.h:
```diff
@@ -133,6 +131,7 @@ class Range final : public AbstractRange, public CanMakeWeakPtr<Range> {
     RangeBoundaryPoint m_start;
     RangeBoundaryPoint m_end;
     bool m_isAssociatedWithSelection { false };
+    bool isLiveRange() const final { return true; }
```
I don't think we should move the declaration of this member function down here.
We usually put all private member variables at the end of class.

jesxilin replied:
Got it, I'll move it back to where it was originally.

jesxilin committed:
CSS Custom Highlights WPT painting-invalidation 003 & 006, painting-iframe 006 failing due to not updating positions correctly
https://bugs.webkit.org/show_bug.cgi?id=259521
rdar://112907577

Reviewed by Richard Robinson and Ryosuke Niwa.

Added check for live range in Document::collectRangeDataFromRegister because
we do not want to skip if a live range is already set.
Live ranges are able to change.
Took out null checks in Document::updateHighlightPositions to support that
live ranges can change and there are already existing null checks before setting positions.
Updated TestExpectations due to three passing tests.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::collectRangeDataFromRegister):
(WebCore::Document::updateHighlightPositions):
* Source/WebCore/dom/Range.h:
* Source/WebCore/dom/StaticRange.h:

Canonical link: https://commits.webkit.org/266349@main
